Add ERC: Smart Contract Delegation (#433)
ethereum/ERCs: ERCS/erc-7710.md

---
eip: 7710

> title: Smart Contract Delegation Interfaces

Reviewer: ```suggestion
title: Smart Contract Delegation
```

> description: Minimal interfaces to enable smart contracts to consistently delegate capabilities to other contracts or EOAs.

Reviewer: ```suggestion
description: Interfaces for consistently delegating capabilities to other contracts or EOAs.
```

Small suggestion here to remove some fluff.

author: Ryan McPeck (@McOso), Dan Finlay (@DanFinlay), Rob Dawson (@rojotek), Derek Chiang (@derekchiang)
discussions-to: https://ethereum-magicians.org/t/towards-more-conversational-wallet-connections-a-proposal-for-the-redeemdelegation-interface/16690
status: Draft
type: Standards Track
category: ERC
created: 2024-05-20

> requires: 1271

Reviewer: ```suggestion
requires: 1271, 7579
```

Line 92 strongly implies that reading 7579 is required to understand the meaning of `mode` and `executionCalldata`.

Author: Have added the require into the latest version.

### Terms

- A **Delegator** is a smart contract that can create a delegation. It implements the `ERC7710` interface, which
  implements ERC-1271.
- A **Delegation Manager** is a singleton smart contract that is responsible for validating delegation authority and
  calling on the *Delegator* to execute an action. It implements the `ER7710Manager` interface.
- A **delegation** is an authority given to another address to perform a specific action.
- A **delegate** is a smart contract, smart contract account, or EOA that has authority to redeem a delegation.
- A **redeemer** is a *delegate* that is using a delegation.

### Overview

#### Redeeming a Delegation

When a delegate wishes to redeem a delegation, they call the `redeemDelegation` function on the Delegation Manager and
pass in the Action they want to execute and the proof of authority (ie delegation) which they are executing on behalf
of. The Delegation Manager then verifies the delegation's validity and, if valid, calls the privileged function on the
Delegator which executes the specified capability on behalf of the Delegator.

> ```mermaid

Reviewer: Our renderer doesn't support mermaid. Please replace this with an image, preferably an SVG.

Author: have updated and added in an SVG in the assets folder for this.

flowchart LR
    redeemer(["redeemer"]) --"redeemDelegation"--> Delegation_Manager(["Delegation Manager"])
    Delegation_Manager -.->|validate delegation w/ Action| Delegation_Manager
    Delegation_Manager --"execute delegated action"--> Delegator(["Delegator"])
    Delegator --"executes CALL using Action"--> Target(["Target"])

    classDef action stroke:#333,stroke-width:2px,stroke-dasharray: 5, 5;
    classDef entity fill:#af,stroke:#333,stroke-width:2px;
    class redeemer,Delegator,Delegation_Manager,Target entity;
```

### Interfaces

#### `ERC7710Manager.sol`

The Delegation Manager MUST implement the `redeemDelegation` which will be responsible for validating the delegation
being invoked, and will then call the delegator to execute the action.

The bytes `_authority` passed in as a parameter to the `redeemDelegation` function contains the authority to execute a
specific action on behalf of the delegating contract.

> The `Action` struct has two fields: `mode` and `executionCalldata`, which are defined precisely as in ERC-7579 (under the "Execution Behavior" section).  Briefly, `mode` encodes the "behavior" of the execution, which could be a single call, a batch call, a delegatecall, and others.  `executionCallData` encodes the data of the execution, which typically includes at least a `target`, a `value`, and a `to` address.

Reviewer: ```suggestion
The `Action` struct has two fields: `mode` and `executionCalldata`, which are defined precisely as in [ERC-7579](./eip-7579.md) (under the "Execution Behavior" section).  Briefly, `mode` encodes the "behavior" of the execution, which could be a single call, a batch call, a delegatecall, and others.  `executionCallData` encodes the data of the execution, which typically includes at least a `target`, a `value`, and a `to` address.
```

  execution behaviors such as batching, delegatecall, and CREATE2 impossible.

To solve the first issue, we decided to remove the requirement for the delegator to implement any specific interface.
Rather, we rely on the Delegation Manager to correctly call the delegator, and we rely on the fact that a delegator would
only create `_authority` for a Delegation Manager that knows how to correctly call it.

To solve the second issue, we decied to adopt the execution interface from ERC-7579, which had to solve a similar problem
within the context of modular smart accounts: defining a standardized execution interface that can support many types of
executions.

## Security Considerations

The introduction of customizable authorization terms requires careful consideration of how authorization data is
structured and interpreted. Potential security risks include the misinterpretation of authorization terms and
unauthorized actions being taken if the interface is not properly implemented. It is recommended that applications
implementing this interface undergo thorough security audits to ensure that authorization terms are handled securely.

> Needs discussion.

Reviewer: ```suggestion
Needs discussion. <!-- TODO -->
```

If you use an HTML-style comment, the linter will make sure you remove it before moving to Review.

Author: comment added.
